torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,268 @@
+#include <queue>
+
+#include <torch/csrc/autograd/functions/accumulate_grad.h>
+#include <torch/csrc/autograd/input_buffer.h>
+#include <torch/csrc/distributed/autograd/context/dist_autograd_container.h>
+#include <torch/csrc/distributed/autograd/engine/dist_engine.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+using torch::autograd::AccumulateGrad;
+using torch::autograd::edge_list;
+using torch::autograd::Engine;
+using torch::autograd::GraphRoot;
+using torch::autograd::GraphTask;
+using torch::autograd::Node;
+using torch::autograd::validate_outputs;
+using torch::autograd::variable_list;
+
+DistEngine::DistEngine()
+    : initializedContextIds_(), engine_(Engine::get_default_engine()) {}
+
+DistEngine& DistEngine::getInstance() {
+  static DistEngine engine;
+  return engine;
+}
+
+void DistEngine::validateRootsAndRetrieveEdges(
+    const variable_list& roots,
+    edge_list& rootEdges,
+    variable_list& grads) {
+  TORCH_CHECK(!roots.empty(), "No tensors provided for gradient computation.");
+  TORCH_INTERNAL_ASSERT(rootEdges.empty());
+  TORCH_INTERNAL_ASSERT(grads.empty());
+
+  // Verify roots are all scalar and require gradients.
+  for (const auto& root : roots) {
+    TORCH_CHECK(
+        root.requires_grad(), "requires_grad not set on: ", root.name());
+    TORCH_CHECK(
+        root.numel() == 1,
+        root.name(),
+        " is not a scalar, all roots need to be scalar");
+    TORCH_CHECK(
+        root.grad_fn(),
+        root.name(),
+        " does not have a valid gradient function.");
+
+    // Compute the root edges and generate the appropriate gradients.
+    rootEdges.push_back(root.gradient_edge());
+    grads.push_back(at::ones_like(root));
+  }
+
+  // Validate rootEdges and grads.
+  validate_outputs(
+      rootEdges, grads, [](const std::string& msg) { return msg; });
+}
+
+void DistEngine::computeDependencies(
+    DistAutogradContext& autogradContext,
+    const edge_list& rootEdges,
+    const variable_list& grads,
+    std::shared_ptr<Node>& graphRoot,
+    edge_list& outputEdges) {
+  // Build the graph task and graph root.
+  auto graphTask = std::unique_ptr<GraphTask>(new GraphTask(
+      /* keep_graph */ false, /* create_graph */ false, /* depth */ 0));
+
+  if (graphRoot == nullptr) {
+    // Create a graphRoot only if needed.
+    graphRoot = std::make_shared<GraphRoot>(rootEdges, grads);
+  }
+
+  // Run BFS to traverse the graph locally. The roots of the graph are
+  // GraphRoot and all send functions for this autograd context.
+  std::unordered_set<Node*> seen;
+  std::queue<Node*> queue;
+  queue.push(static_cast<Node*>(graphRoot.get()));
+
+  auto sendFunctions = autogradContext.sendFunctions();
+
+  // Add all the send functions to the queue as roots.
+  for (const auto& mapEntry : sendFunctions) {
+    // Make sure we don't enqueue the same function twice.
```

Comment:
???

(Is this case happening in the `DistEngine::executeSendFunction` callpath? This seems incredibly fragile...)

Reply:
In this case, can't you just skip passing in the graph root? Then you wouldn't have the duplicate problem.

Reply:
The idea is then you unconditionally call `enqueue_on_cpu` after setting up the overall graph task.